api/submit_result.go

```diff
@@ -23,21 +27,40 @@ func newResultSubmitter(api *API) *resultSubmitter {
 }
 
 // Submit submits results for executionID.
-func (s *resultSubmitter) Submit(executionID string, outputKey string, outputData map[string]interface{}) error {
+func (s *resultSubmitter) Submit(executionID string, outputKey string, outputData []byte) error {
+	exec, err := s.processExecution(executionID, outputKey, outputData)
+	go pubsub.Publish(exec.Service.ResultSubscriptionChannel(), exec)
+	return err
+}
+
+// processExecution processes execution and marks it as complated or failed.
+func (s *resultSubmitter) processExecution(executionID string, outputKey string, outputData []byte) (*execution.Execution, error) {
 	exec, err := s.api.execDB.Find(executionID)
 	if err != nil {
-		return err
+		exec.Failed(err)
+		return nil, err
 	}
+
 	exec.Service, err = service.FromService(exec.Service, service.ContainerOption(s.api.container))
 	if err != nil {
-		return err
+		exec.Failed(err)
+		return nil, err
```

Comment:
It should return exec and not the error.
The error is contained in the the exec and will be publish by the parent function.

Reply:
I made a change to both return the exec and err but it's not idiomatic to return a not-nill type when an err returned. We should improve this in future.